### PR TITLE
Wip/mmeszaros

### PR DESCRIPTION
**Pull request by marcellmeszaros**
PR includes the following:

- Gitignore update
- Relocate UnitTestFramework
- Restructure the SafeguardSessions folder
- Factor out common code (SPS)
- Refactor sps connector specific codes (UnitTestFramework)
-  Remove code duplication from makefile (SPS)

### CODE REVIEW

File: UnitTestFramework/UnitTestFramework.pq [hunk 1]
```diff
@@ -4,7 +4,7 @@
 // Version 1.0.0 was exactly the same as Microsoft's framework. Above this version number,
 // it includes our changes and additions
 [Version = "1.6.3"]
-section UnitTestFrameork;
+section UnitTestFramework;
```
Comment (szszeker) [inline UnitTestFramework/UnitTestFramework.pq:7]: Nice catch!

File: UnitTestFramework/UnitTestFramework.pq
Comment (szszeker) [inline UnitTestFramework/UnitTestFramework.pq:1]: The unit tests in this commit break as the Makefile has not been updated.
I'm asking the others, how should we handle this case?

File: SafeguardSessions/Makefile
Comment (szszeker) [inline SafeguardSessions/Makefile:1]: Small typo in the commit message's first paragraph: a new folders -> new folders.

File: SafeguardSessions/Makefile [hunk 2]
```diff
@@ -4,15 +4,15 @@
...
-TEST_CONNECTOR_DIR=test
+TEST_CONNECTOR_DIR=Test
 
 MAIN_PROJECT_CONFIG=SafeguardSessions.proj
-TEST_PROJECT_CONFIG=$(TEST_CONNECTOR_DIR)\test.proj
+TEST_PROJECT_CONFIG=$(TEST_CONNECTOR_DIR)\SafeguardSessionsTest.proj
 
-MAIN_CONNECTOR=bin\OneIdentitySafeguard.mez
-TEST_FRAMEWORK=$(TEST_CONNECTOR_DIR)\bin\UnitTestFramework.mez
+MAIN_CONNECTOR=Deploy\OneIdentitySafeguard\Bin\OneIdentitySafeguard.mez
```
Comment (szszeker) [inline SafeguardSessions/Makefile:12]: Deploy can be a variable.

File: SafeguardSessions/Makefile [hunk 3]
```diff
@@ -4,15 +4,15 @@
...
+TEST_FRAMEWORK=Deploy\UnitTestFramework\Bin\UnitTestFramework.mez
 
-REPORT_TEMPLATE=assets/template/OneIdentitySafeguard.pbit
+REPORT_TEMPLATE=Tools\Template\OneIdentitySafeguard.pbit
```
Comment (szszeker) [inline SafeguardSessions/Makefile:15]: Tools can be a variable.

File: SafeguardSessions/SafeguardSessions.proj [hunk 4]
```diff
@@ -1,43 +1,71 @@
...
-    <ModulesPath>$(MSBuildProjectDirectory)\modules\</ModulesPath>
-    <RequestPath>$(ModulesPath)\request\</RequestPath>
-    <ErrorPath>$(ModulesPath)\error\</ErrorPath>
-    <SchemaPath>$(ModulesPath)\schema\</SchemaPath>
-    <VersionPath>$(ModulesPath)\version\</VersionPath>
-    <DefinitionsPath>$(Schemapath)\definitions\</DefinitionsPath>
   </PropertyGroup>
-  <ItemGroup>
-    <MezContent Include="SafeguardSessions.pq" />
+  <ItemGroup Label="SafeguardCommon">
```
Comment (szszeker) [inline SafeguardSessions/SafeguardSessions.proj:34]: Common code will be introduced in the the next commit. It should be there.

File: SafeguardCommon/Test/Unit/TestCommonLogger.query.pq
Comment (szszeker) [inline SafeguardCommon/Test/Unit/TestCommonLogger.query.pq:1]: No newline at the end of the file.

File: SafeguardCommon/Test/Unit/TestCommonRequest.query.pq
Comment (szszeker) [inline SafeguardCommon/Test/Unit/TestCommonRequest.query.pq:1]: No newline at the end of the file.

File: SafeguardCommon/Test/Unit/TestCommonRequestErros.query.pq
Comment (szszeker) [inline SafeguardCommon/Test/Unit/TestCommonRequestErros.query.pq:1]: No newline at the end of the file.

File: SafeguardCommon/Test/Unit/TestCommonUtils.query.pq
Comment (szszeker) [inline SafeguardCommon/Test/Unit/TestCommonUtils.query.pq:1]: No newline at the end of the file.

File: SafeguardSessions/Src/Modules/Request/ResponseHandler.pqm
Comment (szszeker) [inline SafeguardSessions/Src/Modules/Request/ResponseHandler.pqm:1]: No newline at the end of the file.

File: SafeguardCommon/Makefile
Comment (szszeker) [inline SafeguardCommon/Makefile:1]: General comment: I don't like the Common prefix in the code. I know that filenames has to be unique in the connector, so it is okay there, but the inner names should be as simple as possible.

Basically imports look like this: Extension.LoadFunction("CommonX", "CommonXs.pqm"). Extension.LoadFunction("X", "CommonXs.pqm") is nicer.

I'm open to discuss.

**Review by szszeker — CHANGES_REQUESTED** [review]

**Review by davedusa — DISMISSED** [review]

File: SafeguardSessions/Tools/PerformanceTesting/Results.xlsx
Comment (davedusa) [inline SafeguardSessions/Tools/PerformanceTesting/Results.xlsx:1]: For me it is a bit strange that we put the Results.xlsx in the Tools folder as I would not consider it as a tool. I think `SafeguardSessions/PerformanceTesting` would be better, but I'm open for other suggestions.

What do You think? @szszeker @asulyi @marcellmeszaros

File: SafeguardSessions/Tools/Template/OneIdentitySafeguard.pbit
Comment (davedusa) [inline SafeguardSessions/Tools/Template/OneIdentitySafeguard.pbit:1]: This also seems strange. Simply just `SafeguardSessions/ReportTemplate/OneIdentitySafeguard.pbit`?
What do You think? @szszeker @asulyi @marcellmeszaros

**Review by davedusa — CHANGES_REQUESTED** [review]

File: UnitTestFramework/UnitTestFramework.pq
Comment (davedusa) [reply UnitTestFramework/UnitTestFramework.pq:1]: I can accept it for now.

**Review by davedusa — CHANGES_REQUESTED** [review]
I write my general comments here + the commit message fix suggestion.

**Commit message**
sps: Factor out common code

in the future SPP connector have been moved -> in the future SPP connector ~~have been moved~~.

**SHORT -> SHOW_ONLY_FAILED_SCENARIOS?**
For a second thought, `SHORT` does not sound too descriptive. We know what it does, but will the SPP developers know? Probably yes after shoot-&-try or reading the code thoroughly.

Can we name it something more descriptive, like above? Others, what's your opinion?

File: SafeguardSessions/Tools/Template/OneIdentitySafeguard.pbit
Comment (asulyi) [reply SafeguardSessions/Tools/Template/OneIdentitySafeguard.pbit:1]: :100:  `Tools` means to me something that provides support functionality e.g. log formatter, json to csv converter.

File: SafeguardSessions/Tools/PerformanceTesting/Results.xlsx
Comment (asulyi) [reply SafeguardSessions/Tools/PerformanceTesting/Results.xlsx:1]: I'd say this is loosely part of the documentation:
`SafeguardSessions/development/PerformanceTesting.xls`
`SafeguardSessions/doc/PerformanceTesting.xls`
`SafeguardSessions/docs/PerformanceTesting.xls`

File: UnitTestFramework/UnitTestFramework.pq
Comment (szszeker) [reply UnitTestFramework/UnitTestFramework.pq:1]: We talked about it. The PR should pass for now.

File: SafeguardSessions/Tools/PerformanceTesting/Results.xlsx
Comment (szszeker) [reply SafeguardSessions/Tools/PerformanceTesting/Results.xlsx:1]: I think performance testing can have its dedicated folder.

**Comment by asulyi** [comment]
> **SHORT -> SHOW_ONLY_FAILED_SCENARIOS?** For a second thought, `SHORT` does not sound too descriptive. We know what it does, but will the SPP developers know? Probably yes after shoot-&-try or reading the code thoroughly.

I'm not against it, since I can see it might be mistaken for something that sets the number of tests or length of runtime. 

But I'd be happier with changing it to something like `OUTPUT=short|long` and any other value causing failure.

Alternatively it might be even better to use an evaluator pwsh script and set the output there instead of hacking the UnitTestFramework.

File: SafeguardSessions/Tools/Template/OneIdentitySafeguard.pbit
Comment (szszeker) [reply SafeguardSessions/Tools/Template/OneIdentitySafeguard.pbit:1]: ReportTemplate sounds nice. If for some reason we want to deliver multiple templates, we can also put them there.

**Comment by szszeker** [comment]
> > **SHORT -> SHOW_ONLY_FAILED_SCENARIOS?** For a second thought, `SHORT` does not sound too descriptive. We know what it does, but will the SPP developers know? Probably yes after shoot-&-try or reading the code thoroughly.
> 
> I'm not against it, since I can see it might be mistaken for something that sets the number of tests or length of runtime.
> 
> But I'd be happier with changing it to something like `OUTPUT=short|long` and any other value causing failure.
> 
> Alternatively it might be even better to use an evaluator pwsh script and set the output there instead of hacking the UnitTestFramework.

For the moment, renaming it is enough.

We have PBI for the UnitTestFramework improvements. The future changes can be discussed when it is on the sprint.

**Comment by asulyi** [comment]
https://github.com/OneIdentity/SafeguardPowerBI/blob/436238721de95c8f3be1bfaa9c7c42a0f2ff7969/UnitTestFramework/UnitTestFramework.pq#L304-L326

I'm not sure `TestResultsContentWithoutStartTime` and `TestFetchInfoContentWithoutStartTime` function are the best place here.

These are used in 
- TestFetchInfo.query.pq,
- TestGetData.query.pq and
- TestFetchSessions.query.pq
files.

So, I think these would better in `SafeguardSessions\Test\Common`.

**Comment by davedusa** [comment]
> I'd be happier with changing it to something like `OUTPUT=short|long` and any other value causing failure.
It sounds better, I agree!

File: .gitignore
[the remark this answers is not in the document]
Comment (marcellmeszaros) [reply .gitignore:1]: ACK.

File: UnitTestFramework/UnitTestFramework.pq (hunk 1 above)
Comment (marcellmeszaros) [reply UnitTestFramework/UnitTestFramework.pq:7]: ACK.

File: SafeguardSessions/Makefile
Comment (marcellmeszaros) [reply SafeguardSessions/Makefile:1]: ACK.

File: SafeguardSessions/Makefile (hunk 2 above)
Comment (marcellmeszaros) [reply SafeguardSessions/Makefile:12]: ACK.

File: SafeguardSessions/Makefile (hunk 3 above)
Comment (marcellmeszaros) [reply SafeguardSessions/Makefile:15]: ACK.

File: SafeguardSessions/SafeguardSessions.proj (hunk 4 above)
Comment (marcellmeszaros) [reply SafeguardSessions/SafeguardSessions.proj:34]: ACK.

File: SafeguardCommon/Test/Unit/TestCommonLogger.query.pq
Comment (marcellmeszaros) [reply SafeguardCommon/Test/Unit/TestCommonLogger.query.pq:1]: ACK.

File: SafeguardCommon/Test/Unit/TestCommonRequestErros.query.pq
Comment (marcellmeszaros) [reply SafeguardCommon/Test/Unit/TestCommonRequestErros.query.pq:1]: ACK.

File: SafeguardCommon/Test/Unit/TestCommonUtils.query.pq
Comment (marcellmeszaros) [reply SafeguardCommon/Test/Unit/TestCommonUtils.query.pq:1]: ACK.

File: SafeguardSessions/Src/Modules/Request/ResponseHandler.pqm
Comment (marcellmeszaros) [reply SafeguardSessions/Src/Modules/Request/ResponseHandler.pqm:1]: ACK.

File: SafeguardCommon/Makefile
Comment (marcellmeszaros) [reply SafeguardCommon/Makefile:1]: ACK.

File: SafeguardSessions/Src/Modules/BaseAuthentication.pqm
[the remark this answers is not in the document]
Comment (marcellmeszaros) [reply SafeguardSessions/Src/Modules/BaseAuthentication.pqm:1]: I agree.

File: SafeguardSessions/Src/Modules/Error/ErrorBase.pqm
[the remark this answers is not in the document]
Comment (marcellmeszaros) [reply SafeguardSessions/Src/Modules/Error/ErrorBase.pqm:1]: ACK.

File: SafeguardSessions/Src/Modules/Logger.pqm
[the remark this answers is not in the document]
Comment (marcellmeszaros) [reply SafeguardSessions/Src/Modules/Logger.pqm:1]: ACK.

File: SafeguardCommon/Src/Modules/Error/CommonRequestErrors.pqm [hunk 5]
```diff
@@ -0,0 +1,57 @@
+let
+    ErrorBase = Extension.ImportModule("CommonErrorBase.pqm"),
+    ErrorLog = Extension.ImportFunction("ErrorLog", "CommonLogger.pqm"),
```
[the remark this answers is not in the document]
Comment (marcellmeszaros) [reply SafeguardCommon/Src/Modules/Error/CommonRequestErrors.pqm:3]: ACK.

File: SafeguardSessions/Src/Modules/Search.pqm
[the remark this answers is not in the document]
Comment (marcellmeszaros) [reply SafeguardSessions/Src/Modules/Search.pqm:1]: ACK.

File: SafeguardSessions/Tools/PerformanceTesting/Results.xlsx
Comment (marcellmeszaros) [reply SafeguardSessions/Tools/PerformanceTesting/Results.xlsx:1]: ACK.

File: SafeguardSessions/Tools/Template/OneIdentitySafeguard.pbit
Comment (marcellmeszaros) [reply SafeguardSessions/Tools/Template/OneIdentitySafeguard.pbit:1]: ACK.

**Comment by szszeker** [comment]
> https://github.com/OneIdentity/SafeguardPowerBI/blob/436238721de95c8f3be1bfaa9c7c42a0f2ff7969/UnitTestFramework/UnitTestFramework.pq#L304-L326
> 
> I'm not sure `TestResultsContentWithoutStartTime` and `TestFetchInfoContentWithoutStartTime` function are the best place here.
> 
> These are used in
> 
>     * TestFetchInfo.query.pq,
> 
>     * TestGetData.query.pq and
> 
>     * TestFetchSessions.query.pq
>       files.
> 
> 
> So, I think these would better in `SafeguardSessions\Test\Common`.

My suggestion is to make it out of scope. Keep the functions of the UnitTestFramework in one place for now.

File: SafeguardSessions/Src/SafeguardSessions.pq
Comment (asulyi) [inline SafeguardSessions/Src/SafeguardSessions.pq:1]: Unused imports:
Request.GetDefaultParameters
Schema.GetSchema
SchemaUtils.ApplySchema

**Review by asulyi — CHANGES_REQUESTED** [review]
I've checked imports of the change files (not the ones that were only renamed though) and found some more, but haven't tested removing them.

Also I've made an attempt fixing case of `SafeguardSessions/test` in 84a7abf6920bd85f7ec04f8e66c91136b0585b66.

File: SafeguardCommon/Src/Modules/Error/CommonRequestErrors.pqm (hunk 5 above)
[the remark this answers is not in the document]
Comment (marcellmeszaros) [reply SafeguardCommon/Src/Modules/Error/CommonRequestErrors.pqm]: Yeah, i somehow forgot it, sorry.

File: SafeguardCommon/Src/Modules/Request/CommonResponseHandler.pqm [hunk 6]
```diff
@@ -1,7 +1,6 @@
 let
-    Logger.ErrorLog = Extension.ImportFunction("ErrorLog", "Logger.pqm"),
-    RequestErrors.ErrorMap = Extension.ImportFunction("ErrorMap", "RequestErrors.pqm"),
-    RequestErrors.NotParsableResponse = Extension.ImportFunction("NotParsableResponse", "RequestErrors.pqm"),
+    Logger.ErrorLog = Extension.ImportFunction("ErrorLog", "CommonLogger.pqm"),
```
[the remark this answers is not in the document]
Comment (marcellmeszaros) [reply SafeguardCommon/Src/Modules/Request/CommonResponseHandler.pqm]: ACK.

File: SafeguardCommon/Test/Unit/TestCommonLogger.query.pq [hunk 7]
```diff
@@ -0,0 +1,95 @@
+section TestCommonLogger;
+
+Logger = Extension.ImportModule("CommonLogger.pqm");
+
+Logger.InfoLog = Logger[InfoLog];
+Logger.ErrorLog = Logger[ErrorLog];
```
[the remark this answers is not in the document]
Comment (marcellmeszaros) [reply SafeguardCommon/Test/Unit/TestCommonLogger.query.pq]: ACK.

File: SafeguardSessions/Src/Modules/Request/Request.pqm [hunk 8]
```diff
@@ -1,6 +1,6 @@
 let
-    IsMocked = Extension.ImportFunction("IsMocked", "Utils.pqm"),
-    InfoLog = Extension.ImportFunction("InfoLog", "Logger.pqm"),
+    IsMocked = Extension.ImportFunction("IsMocked", "CommonUtils.pqm"),
+    InfoLog = Extension.ImportFunction("InfoLog", "CommonLogger.pqm"),
     RequestErrors = Extension.ImportModule("RequestErrors.pqm"),
```
[the remark this answers is not in the document]
Comment (marcellmeszaros) [reply SafeguardSessions/Src/Modules/Request/Request.pqm]: ACK.

File: SafeguardSessions/Src/Modules/Version/Version.pqm [hunk 9]
```diff
@@ -1,10 +1,10 @@
 let
-    IsMocked = Extension.ImportFunction("IsMocked", "Utils.pqm"),
+    IsMocked = Extension.ImportFunction("IsMocked", "CommonUtils.pqm"),
     InfoEndpoint = Extension.ImportFunction("InfoEndpoint", "Constants.pqm"),
-    Logger.InfoLog = Extension.ImportFunction("InfoLog", "Logger.pqm"),
+    Logger.InfoLog = Extension.ImportFunction("InfoLog", "CommonLogger.pqm"),
```
[the remark this answers is not in the document]
Comment (marcellmeszaros) [reply SafeguardSessions/Src/Modules/Version/Version.pqm]: ACK.

File: SafeguardSessions/Src/SafeguardSessions.pq
Comment (marcellmeszaros) [reply SafeguardSessions/Src/SafeguardSessions.pq:1]: ACK.

**Review by asulyi — APPROVED** [review]

File: UnitTestFramework/UnitTestFramework.pq [hunk 10]
```diff
@@ -30,7 +30,7 @@ shared Facts = (_subject as text, _predicates as list) =>
 
 shared Facts.Summarize = (_facts as list) as table =>
     [
-        shortOutput = Text.Lower(Environment.FeatureSwitch("short", "true")) = "true",
```
Comment (szszeker) [inline UnitTestFramework/UnitTestFramework.pq:33]: IMHO, the name can stay shortOutput. It is more descriptive.

File: SafeguardSessions/SafeguardSessions.proj
Comment (szszeker) [inline SafeguardSessions/SafeguardSessions.proj:1]: #Optional: No newline at the end of file.

File: SafeguardCommon/Src/Modules/Error/CommonErrorBase.pqm
Comment (szszeker) [inline SafeguardCommon/Src/Modules/Error/CommonErrorBase.pqm:1]: #Optional: pqm-s don't need a now line at the end.

**Review by szszeker — DISMISSED** [review]

File: SafeguardSessions/Test/Utilities/ConnectorOutputTesters.pqm
Comment (davedusa) [inline SafeguardSessions/Test/Utilities/ConnectorOutputTesters.pqm:1]: Just a nit-pick about the naming. :sweat_smile:  Can we name it to something like `Asserts.pqm` or `Facts.pqm`? IMHO it is more informative and future-proof.

**Review by davedusa — CHANGES_REQUESTED** [review]

File: UnitTestFramework/UnitTestFramework.pq (hunk 10 above)
Comment (marcellmeszaros) [reply UnitTestFramework/UnitTestFramework.pq:33]: ACK.

File: SafeguardSessions/SafeguardSessions.proj
Comment (marcellmeszaros) [reply SafeguardSessions/SafeguardSessions.proj:1]: ACK.

File: SafeguardSessions/Makefile [hunk 11]
```diff
@@ -14,7 +14,7 @@ TEST_FRAMEWORK=$(TEST_CONNECTOR_DIR)\bin\UnitTestFramework.mez
 
 REPORT_TEMPLATE=assets/template/OneIdentitySafeguard.pbit
 
-SHORT?=true
+OUTPUT?=short
```
[the remark this answers is not in the document]
Comment (marcellmeszaros) [reply SafeguardSessions/Makefile:19]: ACK.

File: UnitTestFramework/UnitTestFramework.pq [hunk 12]
```diff
@@ -30,7 +30,7 @@ shared Facts = (_subject as text, _predicates as list) =>
...
+        output = Text.Lower(Environment.FeatureSwitch("output", "short")) = "short",
```
[the remark this answers is not in the document]
Comment (marcellmeszaros) [reply UnitTestFramework/UnitTestFramework.pq:33]: ACK.

File: SafeguardSessions/Test/Utilities/ConnectorOutputTesters.pqm
Comment (marcellmeszaros) [reply SafeguardSessions/Test/Utilities/ConnectorOutputTesters.pqm:1]: ACK.

File: SafeguardCommon/Src/Modules/Error/CommonErrorBase.pqm
Comment (marcellmeszaros) [reply SafeguardCommon/Src/Modules/Error/CommonErrorBase.pqm:1]: ACK.

**Review by davedusa — APPROVED** [review]

**Review by szszeker — APPROVED** [review]

**Pull request merged**